frontend/src/utils/navigate.ts

```diff
@@ -0,0 +1,42 @@
+import type { NavigateFunction, NavigateOptions, To } from 'react-router';
+
+const navigate = (
+	eventName: 'SAFE_NAVIGATE' | 'UNSAFE_NAVIGATE',
+	to: To | number,
+	options?: NavigateOptions,
+): void => {
+	if (window !== undefined) {
+		window.dispatchEvent(
+			new CustomEvent(eventName, {
+				detail: {
+					to,
+					options,
+				},
+				bubbles: false,
+				cancelable: false,
+			}),
+		);
+	} else {
+		throw new Error('Failed navigation from non-component: window is undefined');
+	}
+};
+
+/*
+    Use case:
+        1. Navigate from outside a component
+        2. Inside useEffect, useCallbacks or useMemo to avoid rerendering
+    TODO: Smit need a better name for this
```

Comment:
Are the function name `safeNavigateNoSameURLMemo` and `unsafeNavigateSameURLMemo` is this self-explanatory?